core/src/impl/Kokkos_Core.cpp

```diff
@@ -228,8 +228,20 @@ int get_ctest_gpu(const char* local_rank_str) {
 
 // function to extract gpu # from args
 int get_gpu(const InitArguments& args) {
-  int use_gpu           = args.device_id;
-  const int ndevices    = args.ndevices;
+  int use_gpu        = args.device_id;
+  const int ndevices = args.ndevices > 0 ? args.ndevices :
+#if defined(KOKKOS_ENABLE_CUDA)
+                                         Cuda::detect_device_count();
+#elif defined(KOKKOS_ENABLE_HIP)
+                                         Experimental::HIP::
+                                             detect_device_count();
+#elif defined(KOKKOS_ENABLE_SYCL)
+                                         sycl::device::get_devices(
+                                             sycl::info::device_type::gpu)
+                                             .size();
+#else
+                                         args.ndevices;
+#endif
```

Comment:
This is not very readable.  Try this
```suggestion
  const int ndevices = [](int num_devices) {
#if defined(KOKKOS_ENABLE_CUDA)
    return num_devices > 0 ? num_devices : Cuda::detect_device_count();
#elif defined(KOKKOS_ENABLE_HIP)
    return num_devices > 0 ? num_devices
                           : Experimental::HIP::detect_device_count();
#elif defined(KOKKOS_ENABLE_SYCL)
    return num_devices > 0
               ? num_devices
               : sycl::device::get_devices(sycl::info::device_type::gpu).size();
#else
    return num_devices;
#endif
  }(args.ndevices);
```

Reply:
I used a slight variation.